### PR TITLE
Enhancements for inline volume migration

### PR DESCRIPTION
**Pull request by ddebroy**
Clarifying the enhancements necessary for inline volume migration.

/cc @davidz627 @leakingtapan @msau42

### CODE REVIEW

File: contributors/design-proposals/storage/csi-migration.md [hunk 1]
```diff
@@ -179,6 +179,10 @@ type CSITranslator interface {
   // by the `Driver` field in the CSI Source. The input PV object will not be modified.
   TranslateCSIPVToInTree(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
 
+  // TranslateInTreeVolumeToCSI takes an inline intree volume and will translate
```
Comment (davidz627) [inline contributors/design-proposals/storage/csi-migration.md]: nit: `TranslateInTreeInlineVolumeToCSI`

File: contributors/design-proposals/storage/csi-migration.md [hunk 2]
```diff
@@ -237,15 +245,35 @@ with the in-tree plugin, the VolumeAttachment object becomes orphaned.
...
+The CSI Attacher will have to be modified to also check for `InlineCSIVolumeSource`
+besides the `PersistentVolumeName`. Only one of the two may be specified. If `PersistentVolumeName` 
+is empty and `InlineCSIVolumeSource` is set, the CSI Attacher will not look for 
+an associated PV in it's PV informer cache as it implies the inline volume scenario 
+(where no PVs are created).
+
+The CSI Attacher will have access to most of the data it requires for handling in-line 
+volumes attachment (through the CSI plugins) from `InlineCSIVolumeSource`. However 
+a few fields require special handling due to the absence of a PV where such fields
+are assigned desired values. Specifically, reasonable `MountOptions` and
```
Comment (davidz627) [inline contributors/design-proposals/storage/csi-migration.md]: I guess the question is how are inline volumes handled currently? We want the MountOptions and AccessModes for each plugin to match the default (or chosen?) values/behavior that would have been done without migration.

It is worth calling out in the design that is the philosophy used when choosing the per-plugin defaults.

**Comment by davidz627** [comment]
/assign @jsafrane @saad-ali 
PTAL at this, there is some special handling that needs to be done for inline volumes. We want to make sure it's being done right

**Comment by ddebroy** [comment]
Totally agree with the suggestion to enhance the field in `VolumeAttachment` to be a PV Spec rather than Volume Src. I will reflect the above changes in the doc accordingly.

File: contributors/design-proposals/storage/csi-migration.md [hunk 3]
```diff
@@ -179,6 +179,10 @@ type CSITranslator interface {
   // by the `Driver` field in the CSI Source. The input PV object will not be modified.
   TranslateCSIPVToInTree(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
 
+  // TranslateInTreeVolumeToCSI takes an inline intree volume and will translate
+  // the in-tree volume source to a CSIPersistentVolumeSource
+  // A PV object containing the CSIPersistentVolumeSource in it's spec is returned
+  TranslateInTreeInlineVolumeToCSI(volume *v1.Volume (*v1.PersistentVolume, error) {
```
Comment (jsafrane) [inline contributors/design-proposals/storage/csi-migration.md]: Missing `)` after function parameters.

File: contributors/design-proposals/storage/csi-migration.md (hunk 2 above)
Comment (jsafrane) [reply contributors/design-proposals/storage/csi-migration.md]: In-line volumes don't have `MountOptions`. Volume plugin defaults are always used.

**Comment by liggitt** [comment]
@msau42 - I couldn't tell based on https://github.com/kubernetes/community/pull/3631#issuecomment-487244284... is this making REST-API-visible changes you wanted review for?

**Comment by msau42** [comment]
Ah sorry, the API change is discussed [here](https://github.com/kubernetes/community/pull/3631#pullrequestreview-230422349) but has not been reflected in the kep yet

**Comment by ddebroy** [comment]
@msau42 @davidz627 @saad-ali , I have updated the doc to reflect the change suggested in https://github.com/kubernetes/community/pull/3631#pullrequestreview-230422349

**Comment by davidz627** [comment]
/lgtm
@liggitt PTAL for API changes

**Comment by liggitt** [comment]
> While discussing this @msau42 brought up an interesting question: Do we really need a InlineCSIVolumeSource that is a subset of fields in PV object, or can just use the entire PV object such that the external-attacher only needs to operate on PV objects.

Are the fields in v1.PersistentVolumeSpec other than v1.PersistentVolumeSource relevant for attaching volumes  today? If so, do any of the inline volumes contain data that would translate into those fields? At first glance, the other fields seem to affect mount (AccessModes, MountOptions, VolumeMode), scheduling (Capacity, NodeAffinity), or PV lifecycle (ClaimRef, PersistentVolumeReclaimPolicy, StorageClassName), but not attach.

Would we even allow data in those other PersistentVolumeSpec fields in API validation of a VolumeAttachment?

**Comment by ddebroy** [comment]
@liggitt, yes the CSI external attacher requires multiple fields from the PV Spec - specifically `AccessModes` and `MountOptions` which are combined into a `VolumeCapability` - besides `PersistentVolumeSource`. Here is the call chain that creates `VolumeCapability` in the CSI attacher: [csiAttach](https://github.com/kubernetes-csi/external-attacher/blob/master/pkg/controller/csi_handler.go#L284) => [GetVolumeCapabilities](https://github.com/kubernetes-csi/external-attacher/blob/master/pkg/controller/util.go#L138:6) that is passed down to the CSI plugins.

The fields above will need to be populated by the per-plugin translation logic (as mentioned) for inline volumes. For example:
[1] In case of shared FS plugins like AzureFile/NFS/CephFS, access mode should be set to `csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER`
[2] In case of block plugins like GCE PD, access mode should be set to `csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY` / `csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER` depending on the inline volume source's readonly field.
[3] EBS CSI only supports `csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER` - so the EBS translation logic will always blindly set that.
[4] Appropriate Mount options will need to be set for shared FS like AzureFile.
[5] `ReadOnly` mount option can be set based on a volume source's `ReadOnly` field except for vspherevolume inline volume source where that field [does not exist](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go) and will need to default to RW.

File: contributors/design-proposals/storage/csi-migration.md [hunk 4]
```diff
@@ -236,34 +237,49 @@ with the in-tree plugin, the VolumeAttachment object becomes orphaned.
...
-we will translate the volume source and copy it to the field
-VolumeAttachment.Spec.Source.VolumeAttachmentSource.InlineVolumeSource. The
-VolumeAttachment name must be made with the CSI Translated version of the
+In-line controller calls are a special case because there is no PV. In this case,
+we will translate the in-line Volume into a PersistentVolumeSpec using
+plugin-specific translation logic in the CSI translation library method,
+`TranslateInTreeInlineVolumeToPVSpec`. The resulting PersistentVolumeSpec will
+be stored in a new field `VolumeAttachment.Spec.Source.VolumeAttachmentSource.InlineVolumeSpec`.
+
+The plugin-specific CSI translation logic invoked by `TranslateInTreeInlineVolumeToPVSpec`
```
Comment (msau42) [inline contributors/design-proposals/storage/csi-migration.md]: Can you add a note here, that the PV spec that the translation library produces must be able to pass PV API validation.

Comment (msau42) [reply contributors/design-proposals/storage/csi-migration.md]: And also a note that we need to run the InlineVolumeSpec through the same defaulting and validation functions that we do for PV

Comment (liggitt) [reply contributors/design-proposals/storage/csi-migration.md]: > And also a note that we need to run the InlineVolumeSpec through the same defaulting and validation functions that we do for PV

I think that would happen automatically (but double check the generated `SetDefaults_VolumeAttachment` method) if `SetDefaults_PersistentVolume` was split out to set defaults relevant to all consumers of `PersistentVolumeSpec` (like `VolumeMode`) in a `SetDefaults_PersistentVolumeSpec` function:
```
func SetDefaults_PersistentVolume(obj *v1.PersistentVolume) {
	if obj.Status.Phase == "" {
		obj.Status.Phase = v1.VolumePending
	}
	if obj.Spec.PersistentVolumeReclaimPolicy == "" {
		obj.Spec.PersistentVolumeReclaimPolicy = v1.PersistentVolumeReclaimRetain
	}
	if obj.Spec.VolumeMode == nil {
		obj.Spec.VolumeMode = new(v1.PersistentVolumeMode)
		*obj.Spec.VolumeMode = v1.PersistentVolumeFilesystem
	}
}
```

**Comment by liggitt** [comment]
> @liggitt, yes the CSI external attacher requires multiple fields from the PV Spec - specifically `AccessModes` and `MountOptions` which are combined into a `VolumeCapability` - besides `PersistentVolumeSource`. Here is the call chain that creates `VolumeCapability` in the CSI attacher: [csiAttach](https://github.com/kubernetes-csi/external-attacher/blob/master/pkg/controller/csi_handler.go#L284) => [GetVolumeCapabilities](https://github.com/kubernetes-csi/external-attacher/blob/master/pkg/controller/util.go#L138:6) that is passed down to the CSI plugins.

ok, good to know. in that case, yes, we need all the data required to make the attach call

**Comment by liggitt** [comment]
The API change lgtm.

Can add validation/defaulting details in follow-up if needed.

The last time I looked at this we had `InlineVolumeSource *v1.VolumeSource`. This PR is bumping up a level to `*v1.PersistentVolumeSpec` for reasons I now understand. It looks like in-between, it got changed to `InlineCSIVolumeSource *v1.CSIPersistentVolumeSource` in https://github.com/kubernetes/community/pull/3508, but I can't tell why from that PR's comments.

I just wanted to make sure the reasons we *narrowed* from `*v1.VolumeSource` to `*v1.CSIPersistentVolumeSource` weren't being countermanded by this PR.

File: contributors/design-proposals/storage/csi-migration.md (hunk 4 above)
Comment (liggitt) [reply contributors/design-proposals/storage/csi-migration.md]: > Can you add a note here, that the PV spec that the translation library produces must be able to pass PV API validation.

You can also add tests for this in kubernetes/kubernetes, which has visibility to a) the API types, b) the translation library, and c) the API validation functions

**Comment by davidz627** [comment]
@liggitt I think the reason why we narrowed it to `CSIPersistentVolumeSource` was that we realized that since the `external-attacher` uses a `CSIPersistentVolumeSource` to extract the information it needs to work we could translate the source on the Kubernetes side to send over a type that the attacher can just natively understand. I don't think there was a reason for this change besides a small implementation optimization, it is still possible to do it this way with `PVSpec`

**Comment by msau42** [comment]
Actually I think the reason is that we decided we're not supporting CSIVolumeSource (inline) to do persistent storage use cases.

**Comment by liggitt** [comment]
> @liggitt I think the reason why we narrowed it to `CSIPersistentVolumeSource` was that we realized that since the `external-attacher` uses a `CSIPersistentVolumeSource` to extract the information it needs to work we could translate the source on the Kubernetes side to send over a type that the attacher can just natively understand.

ah, found the conversation in https://github.com/kubernetes/kubernetes/pull/74629#discussion_r261919769 and https://github.com/kubernetes/kubernetes/pull/74629#discussion_r261789835

In my (clearly faulty) memory, we settled on `*v1.PersistentVolumeSource`. If attaching needs content from v1.PersistentVolumeSpec, bumping the type we use up a level is reasonable.

**Pull request merged**